crates/mem-pool/src/pool.rs

```diff
@@ -1181,19 +1208,28 @@ impl MemPool {
         if next_block_number != current_tip_block_number + 1 {
             return Ok(None);
         }
-        let db = self.store.begin_transaction();
-        let tip_block = db.get_last_valid_tip_block()?;
+        let snapshot = self.store.get_snapshot();
+        let tip_block = snapshot.get_last_valid_tip_block()?;
 
+        // update mem block
         let post_merkle_state = tip_block.raw().post_account();
         let mem_block = MemBlock::new(block_info, post_merkle_state);
         self.mem_block = mem_block;
 
         let withdrawals = withdrawals.into_iter().map(Into::into).collect();
-        let mem_store = self.mem_pool_state().load();
-        let mut state = mem_store.state()?;
-        self.finalize_withdrawals(&mut state, withdrawals).await?;
-        self.finalize_deposits(&mut state, deposits).await?;
+        let mem_store = MemStore::new(snapshot);
+        mem_store.update_mem_pool_block_info(self.mem_block.block_info())?;
+        let mut mem_state = mem_store.state()?;
+
+        // remove from pending
+        let db = self.store.begin_transaction();
+        self.remove_unexecutables(&mut mem_state, &db).await?;
+        self.finalize_withdrawals(&mut mem_state, withdrawals)
```

Comment:
`finalize_withdrawals` may fail if read only node cannot collect enough finalized custodians, because e.g. it calls `finalize_withdrawals` later than the full node and some finalized custodian cells are now dead.

Do we want to skip collecting custodian cells in read only nodes?